src/linker/Linker.Steps/ResolveFromXmlStep.cs

```diff
@@ -434,6 +434,7 @@ void MarkMethod (MethodDefinition method)
 				Context.LogMessage ($"Duplicate preserve in {_xmlDocumentLocation} of {method.FullName}"); 
 
 			Annotations.Mark (method);
+			Annotations.MarkReflected (method);
```

Comment:
If a method is private and accessed via a method other than the visible number of callers in the assembly, the call happens through a method other than traditional C#-compiled function calls. This would require either managed reflection or an equivalent unmanaged reflection of the method through the metadata tables.  

If the word "Reflected" portrays a false connotation that this runtime code reflection / invocation happens through the managed System.Reflection, I'm happy to take any suggestions to rename it to. 

I wanted to clarify that it was more specific than just "ExternallyVisible" though. This won't be applied to public methods, only private ones that we know can be accessed.

Reply:
I was thinking of methods called from native code.  It wasn't clear to me if you wanted to capture these as well.

I don't have a better naming suggestion at the moment.

Reply:
> I wanted to clarify that it was more specific than just "ExternallyVisible" though. This won't be applied to public methods, only private ones that we know can be accessed.

Are you also going to prevent the non-reflectable methods from being invoked via reflection (or other indirect means) at runtime?

Otherwise, you can get very weird crashes when the IL linker guesses wrong and the optimizations takes advantage of invariant that does not hold.

We have a very similar attribute for .NET Native, just with a opposite polarity: https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/ReflectionBlockedAttribute.cs. The AOT compiler takes advantage of this attribute for optimizations, but the runtime also prevents these methods from being invoked by reflection.

Reply:
The opposite polarity is nice because it means that files that IL Linker didn't touch (because we didn't run it) will be set up the right way ("assume everything can be accessed by reflection").

Reply:
Oh, nice. I didn't think there already was something out.

I was trying to avoid the size increase of tagging the majority of methods with an attribute. There's definitely soundness gains from doing it the way suggested here though.